### PR TITLE
Add support for handling multiple UwbNotificationData events in simulator driver

### PR DESCRIPTION
**Pull request by abeltrano**
### Type

- [ ] Bug fix
- [X] Feature addition
- [ ] Feature update
- [ ] Breaking change
- [ ] Non-functional change
- [ ] Documentation
- [ ] Infrastructure

### Goals

Allow > 1 simulator clients to listen for notifications.

The notification processing design was overhauled significantly. Previously, there was a global i/o event queue attached to the `UwbSimulatorDevice` instance. This made it difficult to service multiple open-file handles without additional context, and open-file handles are already managed by the WDF framework for which we have a dedicated context object, `UwbSimulatorDeviceFile`.

The design goals are to:
1. Allow each open-file handle to wait on a single `IOCTL_UWB_NOTIFICATION` event.
2. Allow each open-file handle to queue uwb notification data when not actively awaiting on the handle.

The design was updated to have each open-file handle have and manage its own i/o event queue, becoming an i/o event sink. This allows the file handle context to queue notification data when the client is not actively waiting on the `IOCTL_UWB_NOTIFICATION` ioctl, for example, when the client is processing the last event notification data.

#### UwbDeviceIoEventQueue
The i/o event queue, `UwbDeviceIoEventQueue`, is composed of 2 queues: one to queue `UwbNotificationData` entries, called the data queue, and one to queue the single pended `WDFREQUEST` from `DeviceIoControl`, called the request queue. Each queue has a corresponding condition variable associated with it that is used to wait and signal the resource state (request available, notification available). Both condition variables share the same mutex to synchronize access to the shared data and condition variable state. The i/o event queue uses a dedicated thread to wait for the condition variables to be signaled. The high level algorithm is that the processing thread first waits for a request to be pended, then waits for notification data to become available. Once available, it completes the pended `WDFREQUEST`, and repeats until exit is requested.

Since it is expected the client uses the open-file handle for dedicated event processing, the data queue is artificially limited in size to 16 `UwbNotificationData` entries. This may be adjusted in future based on empirical data showing whether that queue size is properly sized.

The `UwbSimulatorDevice` newly exposes functionality to accept uwb notification data, which is then distributed to each open-file handle queue, becoming an i/o event source. This required updating `UwbSimulatorDevice` to track open-file handle contexts (`UwbSimulatorDeviceFile`).

### Technical Details

* Remove i/o event queue from `UwbSimulatorDevice`.
* Track open file handle pointers in `UwbSimulatorDevice`.
* Pend all notification requests unconditionally to simplify processing logic.

### Test Results

* Ran end-to-end ranging scenario and validated state changes and notification data was delivered to `nocli.exe` client.

### Reviewer Focus

* Concurrency here is complicated so it deserves a very close look. In particular, consider corner cases where a deadlock could occur with the condition variables used to wait/signal notification event data and pended request availability in `UwbSimulatorIoEventQueue`.

### Future Work

* Eliminate copies of `UwbNotificationData` being passed around to each open-file i/o event queue.
* Switch to using smart pointers for most objects to manage lifetimes more safely.

### Checklist

- [X] Build target `all` compiles cleanly.
- [X] clang-format and clang-tidy deltas produced no new output.
- [X] Newly added functions include doxygen-style comment block.

### CODE REVIEW

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx [hunk 1]
```diff
@@ -209,11 +172,15 @@ UwbSimulatorDevice::OnFileCreate(WDFDEVICE device, WDFREQUEST request, WDFFILEOB
...
+    auto uwbSimulatorFileBuffer = GetUwbSimulatorDeviceFile(file);
     auto uwbSimulatorFile = new (uwbSimulatorFileBuffer) UwbSimulatorDeviceFile(file, this);
     auto uwbSimulatorFileStatus = uwbSimulatorFile->Initialize();
     if (uwbSimulatorFileStatus == STATUS_SUCCESS) {
         uwbSimulatorFile->RegisterHandler(m_ddiHandler);
+
+        std::unique_lock deviceFilesLockExclusive{ m_deviceFilesGate };
+        m_deviceFiles.push_back(uwbSimulatorFile);
+        DbgPrint("%p added file object %p\n", m_wdfDevice, file);
     } else {
```
Comment (corbin-phipps) [inline windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx:184]: Might be useful to add a log line in the "else" block

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx [hunk 2]
```diff
@@ -209,11 +172,15 @@ UwbSimulatorDevice::OnFileCreate(WDFDEVICE device, WDFREQUEST request, WDFFILEOB
...
         uwbSimulatorFile->~UwbSimulatorDeviceFile();
```
Comment (corbin-phipps) [inline windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx:185]: I don't love this, but it seems like it might be necessary due to the placement new above? Is that necessary because of WDF reasons?

**Review by corbin-phipps — COMMENTED** [review]
Looks good but I want to take another pass later. Lots of concurrency to look at!

File: windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.hxx [hunk 3]
```diff
@@ -58,31 +61,34 @@ public:
...
-     * @brief Handle a WDF request for notification data.
-     *
-     * @param request The WDF request handle.
-     * @param notificationData The output argument to hold notification request data, if available.
-     * @param outputBufferSize The size of the DDI output buffer that will eventually be populated.
-     * @return NTSTATUS
-     */
     NTSTATUS
-    HandleNotificationRequest(WDFREQUEST request, std::optional<::uwb::protocol::fira::UwbNotificationData> &notificationData, std::size_t &outputBufferSize);
+    EnqueueRequest(WDFREQUEST request);
```
Comment (corbin-phipps) [inline windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.hxx:65]: Missing function comment

**Review by corbin-phipps — CHANGES_REQUESTED** [review]

File: windows/drivers/uwb/simulator/UwbSimulatorDeviceFile.hxx [hunk 4]
```diff
@@ -86,6 +90,9 @@ private:
     WDFFILEOBJECT m_wdfFile;
     UwbSimulatorDevice *m_uwbSimulatorDevice{ nullptr };
     std::vector<std::shared_ptr<windows::devices::uwb::simulator::IUwbSimulatorDdiHandler>> m_ddiHandlers{};
+    std::shared_ptr<UwbSimulatorIoEventQueue> m_ioEventQueue;
+
+    static constexpr std::size_t MaximumQueueSizeDefault = 16;
```
[the remark this answers is not in the document]
Comment (corbin-phipps) [reply windows/drivers/uwb/simulator/UwbSimulatorDeviceFile.hxx:109]: Never mind, I see you answered my question in the PR description

**Comment by corbin-phipps** [comment]
Are there unit tests to cover some of the complicated concurrency edge cases? Or are we just validating with simulated ranging data + nocli for now?

File: windows/drivers/uwb/simulator/UwbSimulatorDeviceFile.hxx (hunk 4 above)
[the remark this answers is not in the document]
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorDeviceFile.hxx:109]: I should document it here as well for future refernece.

File: windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.cxx [hunk 5]
```diff
@@ -39,83 +49,108 @@ UwbSimulatorIoEventQueue::Uninitialize()
...
-    auto notificationDataDdi = UwbCxDdi::From(notificationData);
-    auto notificationDataDdiBuffer = std::data(notificationDataDdi);
-    auto notificationDataDdiBufferSize = std::size(notificationDataDdi);
-
-    NTSTATUS status = WdfWaitLockAcquire(m_wdfQueueLock, nullptr);
-    if (!NT_SUCCESS(status)) {
-        return status;
+    // Push the notification data into the queue.
+    {
+        std::unique_lock notificationLockExclusive{ m_notificationGate };
```
[the remark this answers is not in the document]
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.cxx:86]: `std::lock_guard` could be used here, yes. In that case, `std::scoped_lock` is almost always the right choice instead of `std::lock_guard` since it's essentially its successor. I originally used `std::unique_lock` here because the initial implementation used a standard `std::condition_variable` which requires a `std::unique_lock` and I wanted to use consistent RAII locking objects. However, we're now using std::condition_variable_any` which takes any lockable type, not just `std::unique_lock`, so, that should take the `std::scoped_lock` as well.

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx [hunk 6]
```diff
@@ -318,3 +290,16 @@ UwbSimulatorDevice::GetSessionCount()
     std::shared_lock sessionsReadLock{ m_sessionsGate };
     return std::size(m_sessions);
 }
+
+void
+UwbSimulatorDevice::PushUwbNotification(UwbNotificationData uwbNotificationData)
+{
+    // TODO: log
+    // TODO: lock
```
[the remark this answers is not in the document]
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx]: I added a `std::mutex` for this since the first iteration and removed the TODO. 

But to your point: the locking is necessary here, irrespective of what happens in `UwbSimulatorIoEventQueue::PushNotification()` because we're reading `m_deviceFiles` which will be modified any time a handle to the driver is opened, so access to it needs to be synchronized.

However, looking at this again, we really shouldn't continue to hold the lock while executing `PushNotification` for each device file, since we'll already have obtained a `shared_ptr` at that point. I'll make an improvement here.

**Comment by abeltrano** [comment]
> Are there unit tests to cover some of the complicated concurrency edge cases? Or are we just validating with simulated ranging data + nocli for now?

Per our offline conversation, we more or less decided not to develop separate tests for the simulator driver since its purpose is to help with testing the service code. Eventually, the plan is to run the tests for UwbCx against it.

File: windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.hxx (hunk 3 above)
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.hxx:65]: Will update, thanks!

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx (hunk 2 above)
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx:185]: Are you referring to manually invoking the destructor? If so, you're correct that it has to be invoked manually because it was constructed with a contiguous buffer from the WDF framework object, which doesn't have strong type-safety and so won't invoke any kind of destructor on it when it goes out of scope.

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx (hunk 1 above)
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx:184]: Good idea.

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (abeltrano) [reply windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx]: Sorry, my comment is unintelligible. I was actually referring to changes I haven't sent yet which change the pointer type from raw to weak_ptr. I'll make this change on that next PR.

File: windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx (hunk 2 above)
Comment (corbin-phipps) [reply windows/drivers/uwb/simulator/UwbSimulatorDevice.cxx:185]: Yes, that's what I was referring to, and your explanation reinforces my understanding. Thanks!

**Review by corbin-phipps — APPROVED** [review]
